File: clients/ruby/lib/haveapi/client/exceptions.rb
module HaveAPI::Client
  class ProtocolError < StandardError; end

  class ActionFailed < StandardError
    attr_reader :response

    def initialize(response)
      if response.respond_to?(:action)
        msg = response.action.client_message(
          'errors.action_failed',
          action: response.action.name,
          message: response.message
        )
        super(msg)
      else
        super(response.to_s)
      end

      @response = response
    end
  end

  class ValidationError < ActionFailed
    attr_reader :errors

    def initialize(action, errors)
      message = action.client_message('errors.input_parameters_not_valid')
      super(action.client_message('errors.action_failed', action: action.name, message: message))

      @action = action
      @errors = errors
    end
  end
end

File: clients/js/src/haveapi/exceptions.js
/**
 * Thrown when protocol error/incompatibility occurs.
 * @class ProtocolError
 * @memberof HaveAPI.Client.Exceptions
 */
Client.Exceptions.ProtocolError = function (msg) {
	this.name = 'ProtocolError';
	this.message = msg;
}

/**
 * Thrown when calling an action and some arguments are left unresolved.
 * @class UnresolvedArguments
 * @memberof HaveAPI.Client.Exceptions
 */
Client.Exceptions.UnresolvedArguments = function (action) {
	this.name = 'UnresolvedArguments';
	this.message = action.client.translate('errors.unresolved_arguments', {
		action: action.name
	});
}

/**
 * Thrown when trying to cancel an action that cannot be cancelled.
 * @class UncancelableAction
 * @memberof HaveAPI.Client.Exceptions
 */
Client.Exceptions.UncancelableAction = function (stateId, client) {
	this.name = 'UncancelableAction';
	this.message = client.translate('errors.uncancelable_action', {id: stateId});
}
